File: c++/0189-rotate-array.cpp
class Solution {
public:
    // O(n) time, O(1) space - where n is the size of the input array
    void rotate(vector<int>& nums, int k) {
        int n = nums.size();
        k %= n;
        reverse(nums, 0, n - 1);
        reverse(nums, 0, k - 1);
        reverse(nums, k, n - 1);
    }

    // O(n) time, O(1) space - where n is the size of the input array
    void reverse(vector<int>& nums, int start, int end) {
        while (start < end) {
            int temp = nums[start];
            nums[start++] = nums[end];
            nums[end--] = temp; 
        }
    }
};

File: .legacy/python/0189-rotate-array.py
class Solution:
    # O(n) time | O(1) space
    def rotate(self, nums: List[int], k: int) -> None:
        n = len(nums)
        k %= n
        if k == n: return nums
        self.reverse(nums, 0, n - 1)
        self.reverse(nums, 0, k - 1)
        self.reverse(nums, k, n - 1)
        return nums
        
    def reverse(self, nums: List[int], start: int, end: int) -> None:
        while start < end:
            nums[start], nums[end] = nums[end], nums[start]
            start += 1
            end -= 1